src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,856 @@
+      const nodeAsAny = node as any
+      switch (nodeAsAny.type) {
+        case TSNodeType.TSTypeAliasDeclaration:
+          // No typechecking needed, type has already been added to environment
+          return tUndef
+        case TSNodeType.TSAsExpression:
+          const originalType = typeCheckAndReturnType(nodeAsAny.expression, context, env)
+          const typeToCastTo = getAnnotatedType(nodeAsAny, context, env)
+          if ((typeToCastTo as Primitive).name === TSBasicType.ANY) {
+            context.errors.push(new NoExplicitAnyError(nodeAsAny))
+          }
+          if (hasTypeMismatchErrors(typeToCastTo, originalType)) {
+            context.errors.push(
+              new TypecastError(
+                nodeAsAny,
+                formatTypeString(originalType),
+                formatTypeString(typeToCastTo)
+              )
+            )
+          }
+          return typeToCastTo
+        case TSNodeType.TSInterfaceDeclaration:
+          throw new TypeError(node, 'Interface declarations are not allowed')
+        default:
+          throw new TypeError(node, 'Unknown node type')
+      }
+  }
+}
+
+/**
+ * Adds types for imported functions to the type environment.
+ * All imports have their types set to the "any" primitive type.
+ */
+function handleImportDeclarations(
+  node: NodeWithTypeAnnotation<es.Program>,
+  context: Context,
+  env: TypeEnvironment
+) {
+  const importStmts: es.ImportDeclaration[] = node.body.filter(
+    stmt => stmt.type === 'ImportDeclaration'
+  ) as es.ImportDeclaration[]
+  if (importStmts.length > 0) {
+    const modules = memoizedGetModuleManifest()
+    const moduleList = Object.keys(modules)
+    importStmts.forEach(stmt => {
+      const moduleName = stmt.source.value as string
+      if (!moduleList.includes(moduleName)) {
+        context.errors.push(new ModuleNotFoundError(moduleName, stmt))
+      }
+      stmt.specifiers.map(spec => {
+        if (spec.type !== 'ImportSpecifier') {
+          throw new TypeError(stmt, 'Unknown specifier type')
+        }
+
+        setType(spec.local.name, tAny, env)
+      })
+    })
+  }
+}
+
+/**
+ * Adds all types for variable/function/type declarations to the current environment.
+ * This is so that the types can be referenced before the declarations are initialized.
+ */
+function addTypeDeclarationsToEnvironment(
+  node: es.Program | es.BlockStatement,
+  context: Context,
+  env: TypeEnvironment
+) {
+  node.body.forEach(node => {
+    switch (node.type) {
+      case 'FunctionDeclaration':
+        if (node.id === null) {
+          // Block should not be reached since node.id is only null when function declaration
+          // is part of `export default function`, which is not used in Source
+          throw new TypeError(node, 'Function declaration should always have an identifier')
+        }
+        const params = node.params as NodeWithTypeAnnotation<es.Identifier>[]
```

Comment:
`node.params` has type `es.Pattern[]` where `Pattern` is `Identifier | ObjectPattern | ArrayPattern | RestElement | AssignmentPattern | MemberExpression`. Is it correct to typecast a `Pattern` it as just an `Identifier`? Can function arguments be any of the other types? If not, maybe a comment can be added here to explain this.